File: apps/java/design-pattern/src/main/java/pattern/factory/Pond.java
package pattern.factory;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

/**
 * 池（Factory Method パターン）。
 * サブクラスが生成する動物と植物の種類を決定する。
 */
public abstract class Pond {
    private final List<Animal> animals;
    private final List<Plant> plants;

    protected Pond(int numAnimals, Supplier<Animal> animalFactory,
                   int numPlants, Supplier<Plant> plantFactory) {
        animals = new ArrayList<>();
        for (int i = 0; i < numAnimals; i++) {
            animals.add(animalFactory.get());
        }
        plants = new ArrayList<>();
        for (int i = 0; i < numPlants; i++) {
            plants.add(plantFactory.get());
        }
    }

    public List<String> simulateOneDay() {
        List<String> output = new ArrayList<>();
        for (Plant plant : plants) {
            output.add(plant.grow());
        }
        for (Animal animal : animals) {
            output.add(animal.speak());
            output.add(animal.eat());
            output.add(animal.sleep());
        }
        return output;
    }
}

File: apps/ruby/design-pattern/14-factory/lib/pond.rb
require_relative "organisms"

class Pond
  def initialize(number_animals, animal_class, number_plants, plant_class)
    @animals = Array.new(number_animals) { animal_class.new }
    @plants = Array.new(number_plants) { plant_class.new }
  end

  def simulate_one_day
    output = []
    @plants.each { |plant| output << plant.grow }
    @animals.each do |animal|
      output << animal.speak
      output << animal.eat
      output << animal.sleep
    end
    output
  end
end

class DuckPond < Pond
  def initialize(number_animals: 3, number_plants: 2)
    super(number_animals, Duck, number_plants, WaterLily)
  end
end

class FrogPond < Pond
  def initialize(number_animals: 3, number_plants: 2)
    super(number_animals, Frog, number_plants, Algae)
  end
end
